testing/integration/upgrade_downgrade_test.go

```diff
@@ -0,0 +1,49 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+//go:build integration
+
+package integration
+
+import (
+	"context"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	atesting "github.com/elastic/elastic-agent/pkg/testing"
+	"github.com/elastic/elastic-agent/pkg/testing/define"
+	"github.com/elastic/elastic-agent/testing/upgradetest"
+)
+
+func TestStandaloneDowngradeToPreviousSnapshotBuild(t *testing.T) {
+	define.Require(t, define.Requirements{
+		Local: false, // requires Agent installation
+		Sudo:  true,  // requires Agent installation
+	})
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	// Start at the build version as we want to test the retry
+	// logic that is in the build.
+	startFixture, err := define.NewFixture(t, define.Version())
+	require.NoError(t, err)
+
+	// Upgrade to an old build, see `BackwardTwoMinors` for why.
+	upgradeToVersion, err := upgradetest.BackwardTwoMinors(define.Version())
+	require.NoError(t, err)
+	endFixture, err := atesting.NewFixture(
+		t,
+		upgradeToVersion,
+		atesting.WithFetcher(atesting.ArtifactFetcher()),
+	)
+	require.NoError(t, err)
+
+	t.Logf("Testing Elastic Agent upgrade from %s to %s...", define.Version(), upgradeToVersion)
+
+	err = upgradetest.PerformUpgrade(ctx, startFixture, endFixture, t)
+	assert.NoError(t, err)
+}
```

Comment:
This tests was testing the ability to upgrade/downgrade to a specific snapshot build before the refactor... I don't think it's doing that now...

Reply:
Sadly the title of the test did not make that clear. Most test where so hard to follow and read I could not fully understand what the test was doing.

If you can provide me some more detail on what this test is really doing, I can fix it.

Reply:
The original test would select the second-to-last (penultimate) snapshot build from Artifact API, use the buildid to issue a `elastic-agent upgrade x.y.z-SNAPSHOT+<buildid>` in order to distinguish the result from a regular `elastic-agent upgrade x.y.z-SNAPSHOT`

It would then proceed to verify that the commit of the upgraded agent would correspond to the specific build selected, proving that the feature works.

There was also an ugly workaround because the commit hash reported by Artifact API is not always the correct one, but with your approach of creating an endFixture we may sidestep that one if the artifactFetcher is able to pinpoint a specific snapshot build.

Reply:
Okay I will go back and look at the original test and see if there is a clean and simple way of finding the `<buildid>`. The artifact fetcher in the testing library might need to be updated to use that `+<buildid>` as well to pull the correct build.

Reply:
I have updated the test so it is testing what used to be tested. Sorry about that.

I had to add support for the testing framework to work with build ids in the version name as well.